herd/mem.ml

```diff
@@ -508,20 +508,20 @@ module Make(C:Config) (S:Sem.Semantics) : S with module S = S	=
         | Ok ((tgt_proc,code),seen) -> add_code re_exec tgt_proc proc env seen code
 
       and add_fault re_exec inst dir fetch_proc proc env seen addr nexts =
-        match env.A.fh_code with
-        | Some fh_code ->
+        match env.A.fh_code,re_exec with
+        | Some _, true ->
+           Warn.user_error "Cannot handle faults in fault handlers";
+        | Some fh_code, false ->
            add_code true fetch_proc proc env seen fh_code
-        | None -> begin
-            if re_exec then
-              EM.unitcodeT false
-            else
-              let open Precision in
-              match C.precision,dir with
-              | Fatal,_ | LoadsFatal,(Dir.R|Dir.F) -> EM.unitcodeT true
-              | Skip,_ -> add_code false fetch_proc proc env seen nexts
-              | Handled,_ | LoadsFatal,Dir.W ->
-                 add_next_instr true fetch_proc proc env seen addr inst nexts
-          end
+        | None, true ->
+            EM.unitcodeT false
```

Comment:
@maranget, I am not sure I fully understand the significance of the `bool` argument of `unitcodeT`. Do you know how it is used?

Reply:
@relokin Your interpretation of the effect of the `re_exec` parameter is correct (as far as I remember and have checked). The rationale is as follows: in some fault modes, such as the `Handled` mode, executing the faulting instruction is attempted again. There is only one attempt. This behaviour is relevant for tests that, for instance, include code to repair invalid PTE's. Notice that not all (potential) traces are explored, as we miss the traces that correspond to faulting twice or more on the same instruction . Well...